Empire/Empire.cpp

```diff
@@ -151,6 +179,290 @@ void Empire::SetCapitalID(int id) {
+    std::map<std::string, int> total_category_slot_counts = TotalPolicySlots(); // how many slots in each category
+    std::set<std::string> categories_needing_rearrangement;                     // which categories have a problem
+    std::map<std::string, std::map<int, int>> category_slot_policy_counts;      // how many policies in each slot of each category
+    for (auto policy_pair : m_adopted_policies) {
+        const auto& cat = policy_pair.second.category;
+        const auto& slot = policy_pair.second.slot_in_category;
+        const auto& slot_count = category_slot_policy_counts[cat][slot]++;
+        if (slot_count > 1 || policy_pair.second.slot_in_category >= total_category_slot_counts[cat])
+        { categories_needing_rearrangement.insert(cat); }
+    }
+
+    // if a category has too many policies or a slot number conflict, rearrange it
+    // and remove the excess policies
+    for (const auto& cat : categories_needing_rearrangement) {
+        policies_temp = m_adopted_policies;
+
+        // all adopted policies in this category, sorted by slot and adoption turn (lower first)
+        std::multimap<std::pair<int, int>, std::string> slots_turns_policies;
+        for (auto policy_pair : policies_temp) {
+            if (policy_pair.second.category != cat)
+                continue;
+            auto slot = policy_pair.second.slot_in_category;
+            auto turn = policy_pair.second.adoption_turn;
+            slots_turns_policies.insert(std::make_pair(std::make_pair(slot, turn), policy_pair.first));
+            m_adopted_policies.erase(policy_pair.first);    // remove everything from adopted policies in this category...
+        }
+        // re-add in category up to limit, ordered priority by original slot and adoption turn
+        int added = 0;
+        for (auto slot_turn_policy_pair : slots_turns_policies) {
+            if (added >= total_category_slot_counts[cat])
+                break;  // can't add more...
+            m_adopted_policies[slot_turn_policy_pair.second] = {
+                slot_turn_policy_pair.first.second, cat, slot_turn_policy_pair.first.first};
+        }
+    }
+
+    // update counters of how many turns each policy has been adopted
+    for (auto policy_pair : m_adopted_policies)
+        m_policy_adoption_total_duration[policy_pair.first]++;  // assumes default initialization to 0
+
+    // update initial adopted policies for next turn
+    m_initial_adopted_policies = m_adopted_policies;
+}
+
+bool Empire::PolicyAdopted(const std::string& name) const
+{ return m_adopted_policies.count(name); }
+
+int Empire::TurnPolicyAdopted(const std::string& name) const {
+    if (!PolicyAdopted(name))
+        return INVALID_GAME_TURN;
+    auto it = m_adopted_policies.find(name);
+    return it->second.adoption_turn;
+}
+
+int Empire::SlotPolicyAdoptedIn(const std::string& name) const {
+    if (!PolicyAdopted(name))
+        return INVALID_SLOT_INDEX;
+    auto it = m_adopted_policies.find(name);
```

Comment:
When you already check beforehand if the policy exist you can also use the `operator[k]` to cirumvent the iterator and `std::pair<k, v>` access:

```suggestion
    return m_adopted_policies[name].slot_in_category;
```

or even:

```suggestion
    return PolicyAdopted(name) ? m_adopted_policies[name].slot_in_category : INVALID_SLOT_INDEX;
```